ur_controllers/src/passthrough_trajectory_controller.cpp

```diff
@@ -0,0 +1,407 @@
+            "(" +
+            std::to_string(number_of_joints_) + " joint accelerations per point)";
+      RCLCPP_ERROR(get_node()->get_logger(), msg.c_str());
+      msg = "Point nr " + std::to_string(i + 1) +
+            " has: " + std::to_string(goal->trajectory.points[i].accelerations.size()) + " accelerations.";
+      RCLCPP_ERROR(get_node()->get_logger(), msg.c_str());
+      return 0;
+    }
+  }
+  if (goal->trajectory.points[0].velocities.size() == 6 && goal->trajectory.points[0].accelerations.size() == 0) {
+    /*If there are only positions and velocities defined */
+    return 2;
+  } else if (goal->trajectory.points[0].velocities.size() == 0 &&
+             goal->trajectory.points[0].accelerations.size() == 6) {
+    /*If there are only positions and accelerations defined */
+    return 3;
+  } else if (goal->trajectory.points[0].velocities.size() == 6 &&
+             goal->trajectory.points[0].accelerations.size() == 6) {
+    /*If there are both positions, velocities and accelerations defined */
+    return 4;
+  } else {
+    /*If there are only positions defined */
+    return 1;
+  }
+}
+
+// Called when the action is cancelled by the action client.
+rclcpp_action::CancelResponse PassthroughTrajectoryController::goal_cancelled_callback(
+    const std::shared_ptr<
+        rclcpp_action::ServerGoalHandle<control_msgs::action::FollowJointTrajectory>> /* goal_handle */)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Cancelling active trajectory because cancel callback received.");
+  return rclcpp_action::CancelResponse::ACCEPT;
+}
+
+// Action goal was accepted, initialise values for a new trajectory.
+void PassthroughTrajectoryController::goal_accepted_callback(
+    const std::shared_ptr<rclcpp_action::ServerGoalHandle<control_msgs::action::FollowJointTrajectory>> goal_handle)
+{
+  RCLCPP_INFO(get_node()->get_logger(), "Accepted new trajectory.");
+  trajectory_active_ = true;
+  active_trajectory_elapsed_time_ = 0;
+  current_point_ = 0;
+
+  command_interfaces_[CommandInterfaces::PASSTHROUGH_TRAJECTORY_CANCEL].set_value(0.0);
+
+  active_joint_traj_ = goal_handle->get_goal()->trajectory;
+  goal_tolerance_ = goal_handle->get_goal()->goal_tolerance;
+  path_tolerance_ = goal_handle->get_goal()->path_tolerance;
+  goal_time_tolerance_ = goal_handle->get_goal()->goal_time_tolerance;
+  active_goal_ = goal_handle;
+
+  max_trajectory_time_ = active_joint_traj_.points.back().time_from_start.sec +
+                         (active_joint_traj_.points.back().time_from_start.nanosec / pow(10, 9));
+
+  command_interfaces_[CommandInterfaces::PASSTHROUGH_TRAJECTORY_TRANSFER_STATE].set_value(1.0);
+
+  /* Start the executing thread of the action server, this will run until the trajectory is finished or cancelled. */
+  std::thread{ std::bind(&PassthroughTrajectoryController::execute, this, std::placeholders::_1), goal_handle }
+      .detach();
+  return;
+}
+
+void PassthroughTrajectoryController::execute(
```

Comment:
This seems like these checks should be done in the update method, instead, right?